src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -502,9 +499,10 @@ SHAMap::onlyBelow (SHAMapTreeNode* node) const
 
     // An inner node must have at least one leaf
     // below it, unless it's the root_
-    assert (node->hasItem () || (node == root_.get ()));
+    auto leaf = static_cast<SHAMapTreeNode*>(node);
+    assert (leaf->hasItem () || (leaf == root_.get ()));
```

Comment:
`root_` can point to either a `SHAMapInnerNode`, or if the tree only has one item, a `SHAMapTreeNode`.  The type of `root_` is `shared_ptr<SHAMapAbstractNode>` (for this reason).  If `node` points to an inner node (be it `root_` or whatever), then the condition of the `while`: `!node->isLeaf ()` will be false.  And there is no way to exit this while-loop (except by uncaught exception) until `node->isLeaf ()` is true.  This loop-invariant allows the use of `static_cast<SHAMapTreeNode*>(node)` below the loop.